pkg/notifier/notifier.go

```diff
@@ -0,0 +1,176 @@
+package notifier
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"go.uber.org/atomic"
+
+	"github.com/hanfei1991/microcosm/pkg/containers"
+)
+
+type receiverID = int64
+
+// Notifier is the sending endpoint of a single-producer-multiple-consumer
+// notification mechanism.
+type Notifier[T any] struct {
+	receivers sync.Map // receiverID -> *Receiver[T]
+	nextID    atomic.Int64
+
+	queue *containers.SliceQueue[T]
+
+	closeCh       chan struct{}
+	synchronizeCh chan struct{}
+	closeOnce     sync.Once
+}
+
+// Receiver is the receiving endpoint of a single-producer-multiple-consumer
+// notification mechanism.
+type Receiver[T any] struct {
+	id receiverID
+	C  chan T
+
+	closeOnce sync.Once
+	closed    atomic.Bool
+
+	notifier *Notifier[T]
+}
+
+func (r *Receiver[T]) close() {
+	r.closed.Store(true)
+	r.closeOnce.Do(
+		func() {
+			close(r.C)
+		})
+}
+
+// Close closes the receiver
+func (r *Receiver[T]) Close() {
```

Comment:
What is this function used for? Should the receiver be removed from the notifier list when it is closed/removed?

Reply:
My mistake. Fixed in the latest commit.